File: advent2023/01.py

import string
from collections import deque,defaultdict,Counter
from functools import reduce
import re
def lmap(f,l): return list(map(f,l))
def ints(s: str): return lmap(int,re.findall('-?\d+',s))

f = "input01.txt"

with open(f) as r:
    s = r.read() .rstrip()
    lines = s.split('\n')
    lgroups = s.split('\n\n')

def first_last_dig(l, part=1):
    digs = [(x, str(x)) for x in range(10)]
    if part==2: digs += list(enumerate(["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]))
    first = min((l.find(dig), x) for x,dig in digs if dig in l)
    last = max((l.rfind(dig), x) for x,dig in digs if dig in l)
    return 10*first[1] + last[1]

print(sum(first_last_dig(l) for l in lines))
print(sum(first_last_dig(l, part=2) for l in lines))

File: advent2023/01.zig

const std = @import("std");

const input = @embedFile("input01.txt");

const digits: [10][2][]const u8 = .{
    .{"0","zero"},
    .{"1","one"},
    .{"2","two"},
    .{"3","three"},
    .{"4","four"},
    .{"5","five"},
    .{"6","six"},
    .{"7","seven"},
    .{"8","eight"},
    .{"9","nine"},
};

fn first_last(line: []const u8, part: u8) u8 {
    var first_p: usize = line.len+1;
    var last_p: usize = 0;
    var first: u8 = 0;
    var last: u8 = 0;
    for (0.., digits) |i, dig| {
        for(0..part) |p| {
            const f = std.mem.indexOf(u8, line, dig[p]);
            const l = std.mem.lastIndexOf(u8, line, dig[p]);
            if (f != null and f.? < first_p) {
                first_p = f.?;
                first = @intCast(i);
            }
            if (l != null and l.? >= last_p) {
                last_p = l.?;
                last = @intCast(i);
            }
        }
    }
    return first*10 + last;
}

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    var part1: u32 = 0;
    var part2: u32 = 0;
    var it = std.mem.splitScalar(u8, input, '\n');
    while (it.next()) |line| {
        part1 += first_last(line, 1);
        part2 += first_last(line, 2);
    }
    try stdout.print("{}\n{}\n",.{part1, part2});
}